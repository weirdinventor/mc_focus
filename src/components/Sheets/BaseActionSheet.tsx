import React, { useMemo } from 'react';
import { Drawer } from 'vaul';

// -- Helper for dynamic styles --
const Colors = {
  white: '#FFFFFF',
  black1: '#1E1E1E',
  grey: 'rgba(169, 169, 169, 0.5)',
};

// -- Define Prop Types --
type BaseActionSheetProps = {
  open: boolean;
  onOpenChange: (open: boolean) => void;
  children: React.ReactNode;
  snapPoints?: number[];
  initialSnapIndex?: number;
  isLightTheme?: boolean;
};

/**
 * Base ActionSheet component for React JS (`BaseActionSheet`) using Vaul
 *
 * @param {boolean} open - Controls the visibility of the action sheet.
 * @param {function} onOpenChange - Called with the new open state when it changes.
 * @param {Array<number>} [snapPoints] - Snap points ranging from 0 to 1 (fractional). E.g., [0.25, 0.5, 0.9].
 * @param {number} [initialSnapIndex] - Index of the initial snap point from the snapPoints array.
 * @param {React.ReactNode} children - The content of the action sheet.
 * @param {boolean} [isLightTheme=true] - Toggles between light and dark theme.
 * @link
 * https://vaul.emilkowal.ski/
 */
export const BaseActionSheet = ({
  open,
  onOpenChange,
  children,
  snapPoints: snapPointsProp = [1], // Default to 100%
  initialSnapIndex = 0,
  isLightTheme = true,
  ...rest
}: BaseActionSheetProps) => {
  // Memoize active snap point based on the index
  const activeSnapPoint = useMemo(() => {
    return snapPointsProp[initialSnapIndex] || snapPointsProp[0];
  }, [snapPointsProp, initialSnapIndex]);

  // -- Styles --
  // Use React.CSSProperties for strict type checking on styles
  const styles: { [key: string]: React.CSSProperties } = {
    overlay: {
      position: 'fixed',
      inset: '0px',
      backgroundColor: 'rgba(0, 0, 0, 0.55)',
    },
    content: {
      position: 'fixed',
      display: 'flex',
      flexDirection: 'column',
      bottom: '0px',
      left: '0px',
      right: '0px',
      marginTop: '24px',
      borderTopRightRadius: '24px',
      borderTopLeftRadius: '24px',
      paddingBottom: '16px',
      backgroundColor: isLightTheme ? Colors.white : Colors.black1,
      color: isLightTheme ? Colors.black1 : Colors.white,
    },
    indicatorContainer: {
        paddingTop: '12px',
        paddingBottom: '24px',
    },
    indicator: {
      width: '100px',
      height: '4px',
      borderRadius: '99px',
      margin: '0 auto',
      backgroundColor: Colors.grey,
    },
  };

  return (
    <Drawer.Root
      open={open}
      onOpenChange={onOpenChange}
      snapPoints={snapPointsProp}
      activeSnapPoint={activeSnapPoint}
      {...rest}
    >
      <Drawer.Portal>
        <Drawer.Overlay style={styles.overlay} />
        <Drawer.Content style={styles.content}>
            {/* This div structure mimics the original layout with the indicator */}
            <div style={styles.indicatorContainer}>
               <div style={styles.indicator} />
            </div>
            {children}
        </Drawer.Content>
      </Drawer.Portal>
    </Drawer.Root>
  );
};