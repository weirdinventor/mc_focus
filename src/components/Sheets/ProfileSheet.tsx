import { CButton } from './../../components/Buttons/CButton';
import { ControlledInput } from './../../components/ControlledInput';
import { Colors } from './../../constants/Colors';
import { zodResolver } from '@hookform/resolvers/zod';
import {
  useChangeProfilePictureMutation,
  useChangeUsernameMutation,
} from './../../react-query/queries/user/userMutations';
import { useGetMeQuery } from './../../react-query/queries/user/userQueries';
import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import { AiOutlineStar } from 'react-icons/ai';
import { FiEdit2 } from 'react-icons/fi';
import Modal from 'react-modal';
import { UsernameScheme, usernameScheme } from './../../../src/schemes/username.scheme';
import { EditPictureBlock } from './../../modules/profile/components/EditPictureBlock';

// Define the types for the component's props
interface ProfileSheetProps {
  isOpen?: boolean; // Changed to optional
  onRequestClose?: () => void; // Changed to optional
  onClose?: () => void; // Added to align with BaseSheetProps
}

// Set the app element for accessibility
Modal.setAppElement('#root');

export const ProfileSheet = ({ isOpen = false, onRequestClose, onClose }: ProfileSheetProps) => {
  const { data } = useGetMeQuery();
  const { mutate: changePictureMutation } = useChangeProfilePictureMutation();
  const { mutate: changeUsernameMutation } = useChangeUsernameMutation();

  const [cPhoto, setCPhoto] = useState<string | null>(null);
  const [isFocused, setIsFocused] = useState(false);

  const { control, handleSubmit, watch } = useForm<UsernameScheme>({
    defaultValues: { username: data?.username || '' },
    resolver: zodResolver(usernameScheme),
  });

  const watchUsername = watch('username');

  const handleImageChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.files && event.target.files[0]) {
      const file = event.target.files[0];
      const reader = new FileReader();
      reader.onloadend = () => {
        setCPhoto(reader.result as string);
      };
      reader.readAsDataURL(file);
    }
  };

  const onSaveChanges = ({ username }: UsernameScheme) => {
    if (cPhoto) {
      changePictureMutation(cPhoto);
    }
    if (username !== data?.username) {
      changeUsernameMutation({ username });
    }
    onRequestClose?.();
    onClose?.(); // Call onClose if provided
  };

  return (
    <Modal
      isOpen={isOpen}
      onRequestClose={onRequestClose ?? onClose} // Fallback to onClose if onRequestClose is undefined
      contentLabel="Profile Settings"
      style={{
        content: styles.modalContent,
        overlay: styles.overlay,
      }}
    >
      <div style={styles.panel}>
        <EditPictureBlock
          onPress={() => document.getElementById('fileInput')?.click()}
          style={styles.editBlockContainer as any}
          size={88}
          newPicture={cPhoto || undefined}
          iconPosition={{ top: -14, right: -14 }}
        />
        <input
          type="file"
          id="fileInput"
          style={{ display: 'none' }}
          onChange={handleImageChange}
          accept="image/*"
        />
        <ControlledInput
          name="username"
          control={control}
          placeholderText="onboarding.userInfos.usernamePlaceholder"
          onFocus={() => setIsFocused(true)}
          onBlur={() => setIsFocused(false)}
          LeftAccessory={() => (
            <AiOutlineStar
              size={24}
              color={isFocused ? Colors.seance400 : Colors.grey6}
            />
          )}
          RightAccessory={() => (
            <FiEdit2
              size={18}
              color={isFocused ? Colors.seance400 : Colors.grey6}
            />
          )}
        />
        <CButton
          onClick={handleSubmit(onSaveChanges)}
          text="profile.saveModifications"
          buttonType="colored"
          disabled={!(cPhoto || watchUsername !== data?.username)}
        />
      </div>
    </Modal>
  );
};

// Styles remain unchanged
const styles: { [key: string]: React.CSSProperties } = {
  modalContent: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',
    background: '#fff',
    padding: '20px',
    borderRadius: '8px',
    maxWidth: '500px',
    width: '90%',
    border: 'none',
    boxShadow: '0 4px 8px rgba(0, 0, 0, 0.1)',
  },
  overlay: {
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.75)',
    zIndex: 1000,
  },
  panel: {
    paddingLeft: 16,
    paddingRight: 16,
    display: 'flex',
    flexDirection: 'column',
    gap: 24,
    marginTop: 24,
  },
  editBlockContainer: {
    alignSelf: 'center',
  },
};