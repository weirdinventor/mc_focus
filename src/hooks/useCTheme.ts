import { useContext } from 'react';
import { ThemeContext } from './../context/ThemeContext';

export const useCTheme = () => {
  const { theme, toggleTheme } = useContext(ThemeContext);
  const isLightTheme = theme === 'light';

  return { isLightTheme, toggleTheme };
};
