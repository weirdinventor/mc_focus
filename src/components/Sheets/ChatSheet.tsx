import React from 'react';
import type { CSSProperties } from 'react';
import { BaseActionSheet } from './BaseActionSheet';
import { CText } from './../../components/CText';

type ChatSheetProps = {
  open?: boolean;
  onOpenChange?: (isOpen: boolean) => void;
  onClose?: () => void; // Added to align with BaseSheetProps
};

export const ChatSheet = ({ open = false, onOpenChange, onClose }: ChatSheetProps) => {
  return (
    <BaseActionSheet open={open} onOpenChange={onOpenChange ?? (() => onClose?.())}>
      <div style={styles.container}>
        <CText text="Chat Sheet Content" size="lg_bold" />
      </div>
    </BaseActionSheet>
  );
};

const styles: { [key: string]: CSSProperties } = {
  container: {
    paddingLeft: 24,
    paddingRight: 24,
  },
};