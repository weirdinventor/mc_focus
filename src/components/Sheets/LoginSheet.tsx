import React from 'react';
import { useNavigate } from 'react-router-dom';
import IA from './../../assets/images';
import { CButton } from './../../components/Buttons/CButton';
import { CImage } from './../../components/CImage';
import { CText } from './../../components/CText';
import { ProvidersChoices } from './../../modules/onboarding/components/ProvidersChoices';
import { OnboardingStackRoutes, RootStackRoutes } from './../../navigators/routes';
import { openLink } from './../../utils/openLink';
import { BaseActionSheet } from './BaseActionSheet';

type LoginSheetProps = {
  open?: boolean;
  onOpenChange?: (isOpen: boolean) => void;
  onClose?: () => void; // Added to align with BaseSheetProps
};

export const LoginSheet = ({ open = false, onOpenChange, onClose }: LoginSheetProps) => {
  const navigate = useNavigate();

  const onHaveAccountPress = () => {
    onOpenChange?.(false);
    onClose?.(); // Call onClose if provided
    navigate(`${RootStackRoutes.ONBOARDING_STACK}/${OnboardingStackRoutes.LOGIN_SCREEN}`);
  };

  return (
    <BaseActionSheet open={open} onOpenChange={onOpenChange ?? (() => onClose?.())}>
      <div style={styles.container} className='flex flex-col justify-center'>
        <CImage height={72} resizeMode="contain" source={IA.BAG} />
        <CText
          text="onboarding.loginSheet.title"
          size="xxl_extraBold"
          isCentered
          toUppercase
          mb={8}
          mt={24}
        />
        <CText
          text="onboarding.loginSheet.description"
          size="md_medium"
          isCentered
          color="grey"
          mb={24}
        />
        <ProvidersChoices />
        <CText mt={10} isCentered>
          <CText text="onboarding.terms&conditions.agreementText" />
          {' '}
          <CText
            text="onboarding.terms&conditions.terms"
            style={{ textDecoration: 'underline', cursor: 'pointer' }}
            onClick={() =>
              openLink('https://moulaclub.com/app/mobile/politics/cgu')
            }
          />
          {' '}
          <CText text="onboarding.terms&conditions.and" />
          {' '}
          <CText
            text="onboarding.terms&conditions.privacyPolicy"
            style={{ textDecoration: 'underline', cursor: 'pointer' }}
            onClick={() =>
              openLink('https://moulaclub.com/app/mobile/politics/privacy')
            }
          />
        </CText>
        <CButton
          buttonType="secondary"
          text="onboarding.loginSheet.alreadyHave"
          small
          mt={24}
          onClick={onHaveAccountPress}
        />
      </div>
    </BaseActionSheet>
  );
};

const styles: { [key: string]: React.CSSProperties } = {
  container: {
    paddingLeft: 24,
    paddingRight: 24,
  },
};