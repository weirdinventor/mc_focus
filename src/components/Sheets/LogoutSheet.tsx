import React from 'react';
import type { CSSProperties } from 'react';
import { MdLogout } from 'react-icons/md';
import { CText } from './../../components/CText';
import { Colors } from './../../constants/Colors';
import { BaseActionSheet } from './BaseActionSheet';

type LogoutSheetProps = {
  open?: boolean;
  onOpenChange?: (isOpen: boolean) => void;
  onConfirmLogout?: () => void;
  onClose?: () => void; // Added to align with BaseSheetProps
};

export const LogoutSheet = ({
  open = false,
  onOpenChange,
  onConfirmLogout,
  onClose,
}: LogoutSheetProps) => {
  const onDisconnectPress = () => {
    onConfirmLogout?.();
    onOpenChange?.(false);
    onClose?.(); // Call onClose if provided
  };

  return (
    <BaseActionSheet open={open} onOpenChange={onOpenChange ?? (() => onClose?.())}>
      <div style={styles.container}>
        <CText mb={8} size="lg_bold" text="onboarding.logoutSheet.title" />
        <CText size="sm_medium" text="onboarding.logoutSheet.description" />
        <button onClick={onDisconnectPress} style={styles.buttonContainer}>
          <MdLogout color={Colors.deepRed} size={24} />
          <CText
            color="deepRed"
            size="lg_medium"
            text="parameters.items.main.logout"
          />
        </button>
      </div>
    </BaseActionSheet>
  );
};

const styles: { [key: string]: CSSProperties } = {
  container: {
    paddingLeft: 24,
    paddingRight: 24,
  },
  buttonContainer: {
    background: 'none',
    border: 'none',
    padding: 0,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: 24,
    gap: 8,
    cursor: 'pointer',
    width: '100%',
  },
};