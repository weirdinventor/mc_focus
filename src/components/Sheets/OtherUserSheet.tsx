import React from 'react';
import type { CSSProperties } from 'react';
import { useNavigate } from 'react-router-dom';
import { IconicText } from './../../components/IconicText';
import { Colors } from './../../constants/Colors';
import { useBlockUserMutation } from './../../react-query/queries/user/userMutations';
import { BaseActionSheet } from './BaseActionSheet';

type OtherUserSheetProps = {
  open?: boolean;
  onOpenChange?: (isOpen: boolean) => void;
  userId?: string;
  username?: string;
  onClose?: () => void; // Added to align with BaseSheetProps
};

export const OtherUserSheet = ({
  open = false,
  onOpenChange,
  userId,
  username,
  onClose,
}: OtherUserSheetProps) => {
  const { mutate: blockUser } = useBlockUserMutation();
  const navigate = useNavigate();

  const onBlockUser = () => {
    if (userId) {
      blockUser(
        { otherUserId: userId },
        {
          onSuccess: () => {
            onOpenChange?.(false);
            onClose?.(); // Call onClose if provided
            navigate(-1);
          },
        },
      );
    }
  };

  return (
    <BaseActionSheet open={open} onOpenChange={onOpenChange ?? (() => onClose?.())}>
      <div style={styles.container}>
        <button style={styles.buttonStyle} onClick={onBlockUser}>
          <IconicText
            text="onboarding.otherUserSheet.blockUser"
            textOptions={{ username: username ?? '' }} // Fallback for undefined
            textColor="deepRed"
            iconName={'delete'}
          />
        </button>
      </div>
    </BaseActionSheet>
  );
};

const styles: { [key: string]: CSSProperties } = {
  container: {
    paddingLeft: 24,
    paddingRight: 24,
  },
  buttonStyle: {
    background: 'none',
    border: 'none',
    padding: 0,
    margin: 0,
    textAlign: 'left',
    width: '100%',
    cursor: 'pointer',
    paddingTop: 14,
    paddingBottom: 14,
  },
};