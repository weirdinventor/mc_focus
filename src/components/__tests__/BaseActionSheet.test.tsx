import { BaseActionSheet } from '@components/Sheets/BaseActionSheet';
import * as ThemeHook from '@hooks/useCTheme';
import { render } from '@testing-library/react-native';
import React from 'react';
import { Text } from 'react-native';

describe('[COMPONENTS]: BaseActionSheet', () => {
  it('renders correctly with light theme', () => {
    const tree = render(
      <BaseActionSheet sheetId="1">
        <Text>MockedText</Text>
      </BaseActionSheet>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with dark theme', () => {
    jest.spyOn(ThemeHook, 'useCTheme').mockImplementation(() => ({
      isLightTheme: false,
      toggleTheme: () => null,
    }));

    const tree = render(
      <BaseActionSheet sheetId="1">
        <Text>MockedText</Text>
      </BaseActionSheet>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
