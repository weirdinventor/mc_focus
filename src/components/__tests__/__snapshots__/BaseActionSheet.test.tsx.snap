// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[COMPONENTS]: BaseActionSheet renders correctly with dark theme 1`] = `
[
  <RCTSafeAreaView
    collapsable={false}
    onLayout={[Function]}
    pointerEvents="none"
    style={
      {
        "backgroundColor": "transparent",
        "height": 0,
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 1,
      }
    }
  >
    <View />
  </RCTSafeAreaView>,
  <RCTSafeAreaView
    collapsable={false}
    onLayout={[Function]}
    pointerEvents="none"
    style={
      {
        "backgroundColor": "transparent",
        "bottom": 0,
        "height": 1,
        "left": 0,
        "position": "absolute",
        "width": 1,
      }
    }
  >
    <View />
  </RCTSafeAreaView>,
]
`;

exports[`[COMPONENTS]: BaseActionSheet renders correctly with light theme 1`] = `
[
  <RCTSafeAreaView
    collapsable={false}
    onLayout={[Function]}
    pointerEvents="none"
    style={
      {
        "backgroundColor": "transparent",
        "height": 0,
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 1,
      }
    }
  >
    <View />
  </RCTSafeAreaView>,
  <RCTSafeAreaView
    collapsable={false}
    onLayout={[Function]}
    pointerEvents="none"
    style={
      {
        "backgroundColor": "transparent",
        "bottom": 0,
        "height": 1,
        "left": 0,
        "position": "absolute",
        "width": 1,
      }
    }
  >
    <View />
  </RCTSafeAreaView>,
]
`;
